src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -1996,37 +2078,54 @@ public FormValidation doCheckCredentialsId(@CheckForNull @AncestorInPath Item co
                                                                     @QueryParameter String credentialsId) {
+
+            // Start by verifying we have a well-formed GitHub API URL
+            GitHubRepositoryInfo info;
+
+            try {
+                info = GitHubRepositoryInfo.forRepositoryUrl(repositoryUrl);
+            } catch (Throwable e) {
+                if (e instanceof IllegalArgumentException && e.getCause() != null) {
+                    e = e.getCause();
+                }
+                return FormValidation.error(e, e.getMessage());
+            }
+
+            // Do some context permissions checks
             if (context == null && !Jenkins.get().hasPermission(Jenkins.ADMINISTER) ||
                 context != null && !context.hasPermission(Item.EXTENDED_READ)) {
                 return FormValidation.error("Unable to validate repository information"); // not supposed to be seeing this form
             }
             if (context != null && !context.hasPermission(CredentialsProvider.USE_ITEM)) {
                 return FormValidation.error("Unable to validate repository information"); // not permitted to try connecting with these credentials
             }
-            GitHubRepositoryInfo info;
 
-            try {
-                info = GitHubRepositoryInfo.forRepositoryUrl(repositoryUrl);
-            } catch (IllegalArgumentException e) {
-                return FormValidation.error(e, e.getMessage());
+            FormValidation credentialsValidation = Connector.checkScanCredentials(context, info.getApiUri(), credentialsId);
+            if (credentialsValidation.kind == FormValidation.Kind.ERROR) {
+                return credentialsValidation;
             }
 
             StandardCredentials credentials = Connector.lookupScanCredentials(context, info.getApiUri(), credentialsId);
             StringBuilder sb = new StringBuilder();
+            GitHub github;
             try {
-                GitHub github = Connector.connect(info.getApiUri(), credentials);
+                github = Connector.connect(info.getApiUri(), credentials);
                 if (github.isCredentialValid()){
                     sb.append("Credentials ok.");
                 }
+            } catch (IOException e) {
+                return FormValidation.error(e, "Error connecting to API endpoint: " + info.getApiUri());
+            }
 
+            try {
                 GHRepository repo = github.getRepository(info.getRepoOwner() + "/" + info.getRepository());
                 if (repo != null) {
                     sb.append(" Connected to ");
                     sb.append(repo.getHtmlUrl());
                     sb.append(".");
                 }
             } catch (IOException e) {
-                return FormValidation.error(e, "Error validating repository information. " + sb.toString());
+                return FormValidation.error(e, sb.toString() + " Error connecting to repository: " + e.getMessage());
```

Comment:
Is the "successful" information in `sb` in an error case that useful in the UI?

Reply:
The `sb` is built up along the way.  It lets people understand how far verification got before failing.